aten/src/ATen/native/cuda/Unique.cu

```diff
@@ -15,86 +16,103 @@ namespace native{
+  scalar_t *data,
+  int64_t num_inp,
+  const bool return_inverse,
+  const bool return_counts,
+  TensorOptions options,
+  less_t less = thrust::less<scalar_t>(),
+  equal_t equal = thrust::equal_to<scalar_t>()
+) {
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+
+  //sort
+  Tensor inverse_indices;
+  if (!return_inverse) {
+      inverse_indices = at::empty({0}, options);
+      thrust::sort(policy, data, data + num_inp, less);
+  } else {
+      Tensor sorted_indices = at::arange(0, num_inp, options);
+      int64_t* sorted_indices_ptr = sorted_indices.data<int64_t>();
+      thrust::sort_by_key(policy, data, data + num_inp, sorted_indices_ptr, less);
+      Tensor inv_loc = at::empty({num_inp}, options);
+      inverse_indices = at::empty({num_inp}, options);
+      int64_t* inv_loc_ptr = inv_loc.data<int64_t>();
+      int64_t* inverse_indices_ptr = inverse_indices.data<int64_t>();
+      thrust::adjacent_difference(policy, data, data + num_inp, inv_loc_ptr, [=] __device__ (scalar_t a, scalar_t b) -> int64_t { if (a != b) {return 1;} else { return 0; }});
+      inv_loc[0] = 0;
+      thrust::inclusive_scan(policy, inv_loc_ptr, inv_loc_ptr + num_inp, inv_loc_ptr);
+      thrust::scatter(policy, inv_loc_ptr, inv_loc_ptr + num_inp, sorted_indices_ptr, inverse_indices_ptr);
+  }
+
+  // unique
+  Tensor counts = at::empty({0}, options);
+  int64_t num_out;
+  if (!return_counts) {
+      num_out = thrust::unique(policy, data, data + num_inp, equal) - data;
+  } else {
+      Tensor sorted_indices = at::arange(0, num_inp + 1, options);
+      int64_t* sorted_indices_ptr = sorted_indices.data<int64_t>();
+      num_out = thrust::unique_by_key(policy, data, data + num_inp, sorted_indices_ptr, equal).first - data;
+      sorted_indices[num_out] = num_inp;
+      counts.resize_(num_out);
+      int64_t* counts_ptr = counts.data<int64_t>();
+      thrust::adjacent_difference(policy, sorted_indices_ptr + 1, sorted_indices_ptr + num_out + 1, counts_ptr);
+  }
+
+  THCudaCheck(cudaGetLastError());
+  return std::tuple<Tensor, Tensor, int64_t>(inverse_indices, counts, num_out);
+}
+
 template <typename scalar_t>
-  std::tuple<Tensor, Tensor, Tensor> _unique_cuda_template(
+std::tuple<Tensor, Tensor, Tensor> unique_cuda_template(
     const Tensor& self,
     const bool return_inverse,
     const bool return_counts) {
 
-    cudaStream_t stream = at::cuda::getCurrentCUDAStream();
-    auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
-    auto policy = thrust::cuda::par(allocator).on(stream);
-
-    const Tensor& input = self.contiguous();
-    int64_t num_inp = input.numel();
-    const scalar_t* input_data = input.data<scalar_t>();
-
-    //sort
-    Tensor output = input.clone();
-    output = output.view(-1);
+    Tensor output = self.contiguous().reshape(-1).clone();
```

Comment:
nit: there might be unnecessary copies here.
I believe you can get the same behavior with
```python
self.clone().reshape(-1)
```

Reply:
fixed